source/extensions/filters/network/ratelimit/ratelimit.h

```diff
@@ -83,13 +85,44 @@ class Filter : public Network::ReadFilter,
   void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
     filter_callbacks_ = &callbacks;
     filter_callbacks_->connection().addConnectionCallbacks(*this);
+    applySubstitutionFormatter(config_->descriptors(),
+                               filter_callbacks_->connection().streamInfo());
+  }
+
+  void applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
```

Comment:
nit: after refactor per below, move these function bodies into cc files, there is no need to have them in the header files.